code/modules/antagonists/changeling/powers/swap_form.dm

```diff
@@ -44,36 +44,25 @@
 
 	to_chat(target, "<span class='userdanger'>[user] tightens [user.p_their()] grip as a painful sensation invades your body.</span>")
 
-	var/lingpowers = list()
-	for(var/power in changeling.purchasedpowers)
-		lingpowers += power
+	var/datum/dna/DNA = cling.get_dna(user.dna)
+	cling.absorbed_dna -= DNA
+	cling.protected_dna -= DNA
+	cling.absorbed_count--
+	if(!cling.get_dna(target.dna))
+		cling.absorb_dna(target)
+	cling.trim_dna()
 
-	changeling.absorbed_dna -= changeling.find_dna(user.dna)
-	changeling.protected_dna -= changeling.find_dna(user.dna)
-	changeling.absorbedcount -= 1
-	if(!changeling.has_dna(target.dna))
-		changeling.absorb_dna(target, user)
-	changeling.trim_dna()
-
-	var/mob/dead/observer/ghost = target.ghostize(0)
+	var/mob/dead/observer/ghost = target.ghostize(FALSE)
 	user.mind.transfer_to(target)
 	if(ghost && ghost.mind)
 		ghost.mind.transfer_to(user)
 		GLOB.non_respawnable_keys -= ghost.ckey //they have a new body, let them be able to re-enter their corpse if they die
 		user.key = ghost.key
 	qdel(ghost)
 	user.Paralyse(4 SECONDS)
-	target.add_language("Changeling")
-	user.remove_language("Changeling")
 	user.regenerate_icons()
 	if(target.stat == DEAD && target.suiciding)  //If Target committed suicide, unset flag for User
-		target.suiciding = 0
-
-	for(var/power in lingpowers)
-		var/datum/action/changeling/S = power
-		target.mind.changeling.purchasedpowers += S
-		if(istype(S) && S.needs_button)
-			S.Grant(target)
```

Comment:
assuming all of this stuff is handled by `user.mind.transfer_to(target)`

Reply:
Yep `transfer_to`, and by extension `apply_innate_effects()` and `remove_innate_effects`.